File: ios/CookieManagerError.swift
import Foundation

enum CookieManagerErrorCode: String, CaseIterable {
  case invalidURL = "invalid_url"
  case invalidCookie = "invalid_cookie"
  case domainMismatch = "domain_mismatch"
  case notSupported = "not_supported"
  case storageError = "storage_error"
  case networkError = "network_error"
}

struct CookieManagerInputError: LocalizedError {
  let code: CookieManagerErrorCode
  let message: String
  let underlyingError: Error?

  init(
    code: CookieManagerErrorCode,
    message: String,
    underlyingError: Error? = nil
  ) {
    self.code = code
    self.message = message
    self.underlyingError = underlyingError
  }

  var errorDescription: String? { message }
}

File: android/src/main/java/com/preeternal/reactnativecookiemanager/CookieManagerError.kt
package com.preeternal.reactnativecookiemanager

internal enum class CookieManagerErrorCode(val value: String) {
  INVALID_URL("invalid_url"),
  INVALID_COOKIE("invalid_cookie"),
  DOMAIN_MISMATCH("domain_mismatch"),
  NOT_SUPPORTED("not_supported"),
  STORAGE_ERROR("storage_error"),
  NETWORK_ERROR("network_error")
}

internal class CookieManagerException(
  val code: CookieManagerErrorCode,
  message: String,
  cause: Throwable? = null
) : Exception(message, cause)

internal fun cookieManagerErrorCode(
  error: Throwable,
  fallback: CookieManagerErrorCode
): String = (error as? CookieManagerException)?.code?.value ?: fallback.value
